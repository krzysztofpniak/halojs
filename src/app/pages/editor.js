import React from 'react';
import {connect, H, mkComponent, mkEval} from '../../halo/halo';
import Header from '../components/header';
import {Routes} from '../../data/routes';
import {mergeLeft, prop} from 'ramda';
import Footer from '../../footer';
import daggy from 'daggy';
import Form from '../../halo/form';
import fieldTypes from '../form/fieldTypes';
import {A} from '../appM';
import go from '../../halo/helpers/go';
import Maybe from '../../halo/helpers/maybe';
import {unit} from '../../halo/helpers/fns';

const Action = daggy.taggedSum('Action', {
  Initialize: [],
  Receive: ['input'],
  Form: ['formAction'],
});

const handleAction = action =>
  action.cata({
    Initialize: () =>
      go(function* () {
        const {slug} = yield H.get();
        const article = yield slug.cata({
          Just: x =>
            A.getArticle(x).chain(x =>
              x.cata({
                NotAsked: z => Maybe.Nothing,
                Loading: z => Maybe.Nothing,
                Failure: z => Maybe.Nothing,
                Success: article => Maybe.Just(article),
              })
            ),
          Nothing: () => H.of(unit),
        });

        return H.query('Form')('form')(Form.Query.InitFields(article));
      }),
    Receive: input => H.modify(mergeLeft(input)),
    Form: formAction =>
      formAction.cata({
        Submit: fields =>
          go(function* () {
            const {slug} = yield H.get();
            const z = yield slug.cata({
              Nothing: () => A.createArticle(fields),
              Just: s => A.updateArticle(s, fields),
            });
            return z.cata({
              NotAsked: z => H.of(unit),
              Loading: z => H.of(unit),
              Failure: z => H.of(unit),
              Success: ({slug}) => H.navigate(Routes.EditArticle(slug)),
            });
          }),
      }),
  });

const schema = [
  {id: 'title', placeholder: 'Article Title'},
  {id: 'description', placeholder: "What's this article about?"},
  {
    id: 'body',
    type: 'textarea',
    placeholder: 'Write your article (in markdown)',
  },
  {id: 'tagList', placeholder: 'Article Title'},
];

const evalFn = mkEval({
  initialize: Action.Initialize,
  handleAction,
  receive: Action.Receive,
});

const Editor = connect(s => s)(
  mkComponent({
    actionType: Action,
    initialState: ({currentUser}) => ({
      currentUser,
    }),
    evalFn,
  })(({currentUser, Slot, slug}) => (
    <div>
      <Header currentUser={currentUser} route={Routes.Editor} />
      <div className="editor-page">
        <div className="container page">
          <div className="row">
            <div className="col-md-10 offset-md-1 col-xs-12">
              <Slot
                component={Form}
                output={Action.Form}
                slotType="Form"
                slotId="form"
                schema={schema}
                fieldTypes={fieldTypes}
              />
            </div>
          </div>
        </div>
      </div>
      <Footer />
    </div>
  ))
);

export default Editor;

const editorEvalFn = evalFn;

export {editorEvalFn};
