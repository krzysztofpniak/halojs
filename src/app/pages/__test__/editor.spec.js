import {editorEvalFn} from '../editor';
import Pair from '../../../halo/helpers/pair';
import Maybe from '../../../halo/helpers/maybe';
import setupEval from '../../../halo/testing/setupEval';
import renderer from 'react-test-renderer';

const Z = () => <div className="as">dupa</div>;

describe('editorEvalFn', () => {
  it('should asd', () => {
    const {initialize, putState, getState} = setupEval(editorEvalFn);
    putState({slug: Maybe.Just('sdf')});

    expect(initialize()).toEqual(
      Pair(
        [
          'State',
          'GetArticle',
          'ChildQuery: Form.form.Query.InitFields({"a": "hello"})',
        ],
        ''
      )
    );
    expect(getState()).toEqual({slug: Maybe.Just('sdf')});
  });
  it('renders correctly', () => {
    const tree = renderer.create(<Z />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
