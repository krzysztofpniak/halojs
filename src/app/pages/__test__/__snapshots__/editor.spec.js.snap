// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editorEvalFn renders correctly 1`] = `
<div>
  dupa
</div>
`;
